File: r3e-deno/src/js/oracle.js
// Copyright @ 2023 - 2024, R3E Network
// All Rights Reserved

// Oracle service JavaScript API

/**
 * Oracle service for accessing external data
 */
class Oracle {
  /**
   * Submit a generic oracle request
   * @param {string} requestType - Type of oracle request (price, random, weather, sports, custom)
   * @param {Object} data - Request data
   * @param {string} requesterId - Requester ID
   * @param {string} [callbackUrl] - Optional callback URL
   * @returns {Promise<string>} Request ID
   */
  static async submitRequest(requestType, data, requesterId, callbackUrl = null) {
    const config = {
      request_type: requestType,
      data,
      callback_url: callbackUrl,
      requester_id: requesterId,
    };
    
    const result = Deno.core.ops.op_oracle_submit_request(config);
    return result.request_id;
  }
  
  /**
   * Get the status of an oracle request
   * @param {string} requestId - Request ID
   * @returns {Promise<string>} Request status (pending, processing, completed, failed)
   */
  static async getRequestStatus(requestId) {
    const result = Deno.core.ops.op_oracle_get_request_status(requestId);
    return result.status;
  }
  
  /**
   * Get the response for a completed oracle request
   * @param {string} requestId - Request ID
   * @returns {Promise<Object>} Oracle response
   */
  static async getResponse(requestId) {
    return Deno.core.ops.op_oracle_get_response(requestId);
  }
  
  /**
   * Cancel an oracle request
   * @param {string} requestId - Request ID
   * @returns {Promise<boolean>} Success flag
   */
  static async cancelRequest(requestId) {
    const result = Deno.core.ops.op_oracle_cancel_request(requestId);
    return result.success;
  }
  
  /**
   * Get price data for a cryptocurrency
   * @param {string} symbol - Asset symbol (e.g., "NEO", "GAS")
   * @param {string} [currency="USD"] - Currency to convert to
   * @param {string[]} [sources=[]] - Preferred price sources
   * @param {string} requesterId - Requester ID
   * @returns {Promise<string>} Request ID
   */
  static async getPrice(symbol, currency = "USD", sources = [], requesterId) {
    const config = {
      symbol,
      currency,
      sources,
      requester_id: requesterId,
    };
    
    const result = Deno.core.ops.op_oracle_get_price(config);
    return result.request_id;
  }
  
  /**
   * Get price data by index
   * @param {number} index - Price index (0 for NEO/USD, 1 for GAS/USD, 2 for BTC/USD, 3 for ETH/USD)
   * @param {string[]} [sources=[]] - Preferred price sources
   * @param {string} requesterId - Requester ID
   * @returns {Promise<string>} Request ID
   */
  static async getPriceByIndex(index, sources = [], requesterId) {
    const config = {
      index,
      sources,
      requester_id: requesterId,
    };
    
    const result = Deno.core.ops.op_oracle_get_price_by_index(config);
    return result.request_id;
  }
  
  /**
   * Update price data on the blockchain
   * @param {string} symbol - Asset symbol (e.g., "NEO", "GAS")
   * @param {string} requesterId - Requester ID
   * @returns {Promise<string>} Transaction hash
   */
  static async updatePriceOnBlockchain(symbol, requesterId) {
    const config = {
      symbol,
      requester_id: requesterId,
    };
    
    const result = Deno.core.ops.op_oracle_update_price_on_blockchain(config);
    return result.tx_hash;
  }
  
  /**
   * Get random numbers
   * @param {Object} options - Random number options
   * @param {number} [options.min=0] - Minimum value (inclusive)
   * @param {number} [options.max=Number.MAX_SAFE_INTEGER] - Maximum value (inclusive)
   * @param {number} [options.count=1] - Number of random values to generate
   * @param {string} [options.method="secure"] - Random number generation method (secure, blockchain, vrf)
   * @param {string} [options.seed] - Optional seed for deterministic generation
   * @param {string} requesterId - Requester ID
   * @returns {Promise<string>} Request ID
   */
  static async getRandom(options, requesterId) {
    const config = {
      min: options.min,
      max: options.max,
      count: options.count,
      method: options.method,
      seed: options.seed,
      requester_id: requesterId,
    };
    
    const result = Deno.core.ops.op_oracle_get_random(config);
    return result.request_id;
  }
  
  /**
   * Wait for an oracle request to complete
   * @param {string} requestId - Request ID
   * @param {number} [timeout=30000] - Timeout in milliseconds
   * @param {number} [interval=1000] - Polling interval in milliseconds
   * @returns {Promise<Object>} Oracle response
   */
  static async waitForResponse(requestId, timeout = 30000, interval = 1000) {
    const startTime = Date.now();
    
    while (Date.now() - startTime < timeout) {
      const status = await Oracle.getRequestStatus(requestId);
      
      if (status === "completed") {
        return await Oracle.getResponse(requestId);
      } else if (status === "failed") {
        const response = await Oracle.getResponse(requestId);
        throw new Error(`Oracle request failed: ${response.error || "Unknown error"}`);
      }
      
      // Wait for the next polling interval
      await new Promise(resolve => setTimeout(resolve, interval));
    }
    
    throw new Error(`Oracle request timed out after ${timeout}ms`);
  }
  
  /**
   * Get Neo price data
   * @param {string} [currency="USD"] - Currency to convert to
   * @param {string[]} [sources=[]] - Preferred price sources
   * @param {string} requesterId - Requester ID
   * @returns {Promise<Object>} Price data
   */
  static async getNeoPrice(currency = "USD", sources = [], requesterId) {
    const requestId = await Oracle.getPrice("NEO", currency, sources, requesterId);
    return await Oracle.waitForResponse(requestId);
  }
  
  /**
   * Get GAS price data
   * @param {string} [currency="USD"] - Currency to convert to
   * @param {string[]} [sources=[]] - Preferred price sources
   * @param {string} requesterId - Requester ID
   * @returns {Promise<Object>} Price data
   */
  static async getGasPrice(currency = "USD", sources = [], requesterId) {
    const requestId = await Oracle.getPrice("GAS", currency, sources, requesterId);
    return await Oracle.waitForResponse(requestId);
  }
  
  /**
   * Get price data by index from the blockchain
   * @param {number} index - Price index (0 for NEO/USD, 1 for GAS/USD, 2 for BTC/USD, 3 for ETH/USD)
   * @param {string} requesterId - Requester ID
   * @returns {Promise<Object>} Price data
   */
  static async getPriceFromBlockchain(index, requesterId) {
    const config = {
      index,
      requester_id: requesterId,
    };
    
    return Deno.core.ops.op_oracle_get_price_from_blockchain(config);
  }
  
  /**
   * Update price data on the blockchain
   * @param {string} symbol - Asset symbol (e.g., "NEO", "GAS")
   * @param {string} requesterId - Requester ID
   * @returns {Promise<Object>} Transaction result
   */
  static async updatePriceOnBlockchain(symbol, requesterId) {
    const config = {
      symbol,
      requester_id: requesterId,
    };
    
    return Deno.core.ops.op_oracle_update_price_on_blockchain(config);
  }
  
  /**
   * Update price data by index on the blockchain
   * @param {number} index - Price index (0 for NEO/USD, 1 for GAS/USD, 2 for BTC/USD, 3 for ETH/USD)
   * @param {string} requesterId - Requester ID
   * @returns {Promise<Object>} Transaction result
   */
  static async updatePriceByIndexOnBlockchain(index, requesterId) {
    const config = {
      index,
      requester_id: requesterId,
    };
    
    return Deno.core.ops.op_oracle_update_price_by_index_on_blockchain(config);
  }
  
  /**
   * Get a secure random number
   * @param {number} [min=0] - Minimum value (inclusive)
   * @param {number} [max=Number.MAX_SAFE_INTEGER] - Maximum value (inclusive)
   * @param {string} requesterId - Requester ID
   * @returns {Promise<number>} Random number
   */
  static async getSecureRandom(min = 0, max = Number.MAX_SAFE_INTEGER, requesterId) {
    const requestId = await Oracle.getRandom({
      min,
      max,
      count: 1,
      method: "secure",
    }, requesterId);
    
    const response = await Oracle.waitForResponse(requestId);
    return response.data.values[0];
  }
  
  /**
   * Get a blockchain-based random number
   * @param {number} [min=0] - Minimum value (inclusive)
   * @param {number} [max=Number.MAX_SAFE_INTEGER] - Maximum value (inclusive)
   * @param {string} requesterId - Requester ID
   * @returns {Promise<Object>} Random number with proof
   */
  static async getBlockchainRandom(min = 0, max = Number.MAX_SAFE_INTEGER, requesterId) {
    const requestId = await Oracle.getRandom({
      min,
      max,
      count: 1,
      method: "blockchain",
    }, requesterId);
    
    const response = await Oracle.waitForResponse(requestId);
    return {
      value: response.data.values[0],
      proof: response.data.proof,
    };
  }
}

export { Oracle };

File: r3e-deno/src/ext/oracle.rs
// Copyright @ 2023 - 2024, R3E Network
// All Rights Reserved

use deno_core::error::AnyError;
use deno_core::op2;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

use r3e_oracle::service::create_oracle_request;
use r3e_oracle::types::{PriceRequest, PriceResponse, RandomMethod, RandomRequest, RandomResponse};
use r3e_oracle::{
    OracleError, OracleRequest, OracleRequestStatus, OracleRequestType, OracleResponse,
    OracleService,
};

// Oracle request operations

#[derive(Debug, Serialize, Deserialize)]
pub struct OracleRequestConfig {
    pub request_type: String,
    pub data: serde_json::Value,
    pub callback_url: Option<String>,
    pub requester_id: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OracleRequestResult {
    pub request_id: String,
}

#[op2]
#[serde]
pub fn op_oracle_submit_request(
    #[serde] config: OracleRequestConfig,
    #[state] oracle_service: &Arc<dyn OracleService>,
) -> Result<OracleRequestResult, AnyError> {
    // Convert request type string to enum
    let request_type = match config.request_type.as_str() {
        "price" => OracleRequestType::Price,
        "random" => OracleRequestType::Random,
        "weather" => OracleRequestType::Weather,
        "sports" => OracleRequestType::Sports,
        "custom" => OracleRequestType::Custom,
        _ => {
            return Err(AnyError::msg(format!(
                "Unsupported request type: {}",
                config.request_type
            )))
        }
    };

    // Convert data to string
    let data = serde_json::to_string(&config.data)
        .map_err(|e| AnyError::msg(format!("Failed to serialize request data: {}", e)))?;

    // Create oracle request
    let request =
        create_oracle_request(request_type, data, config.callback_url, config.requester_id);

    // Store request ID for response
    let request_id = request.id.clone();

    // Submit request
    let rt = tokio::runtime::Runtime::new().unwrap();
    rt.block_on(async {
        oracle_service
            .submit_request(request)
            .await
            .map_err(|e| AnyError::msg(format!("Failed to submit request: {}", e)))
    })?;

    Ok(OracleRequestResult { request_id })
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OracleStatusResult {
    pub status: String,
}

#[op2]
#[serde]
pub fn op_oracle_get_request_status(
    #[string] request_id: String,
    #[state] oracle_service: &Arc<dyn OracleService>,
) -> Result<OracleStatusResult, AnyError> {
    // Get request status
    let rt = tokio::runtime::Runtime::new().unwrap();
    let status = rt.block_on(async {
        oracle_service
            .get_request_status(&request_id)
            .await
            .map_err(|e| AnyError::msg(format!("Failed to get request status: {}", e)))
    })?;

    // Convert status to string
    let status_str = match status {
        OracleRequestStatus::Pending => "pending",
        OracleRequestStatus::Processing => "processing",
        OracleRequestStatus::Completed => "completed",
        OracleRequestStatus::Failed => "failed",
    };

    Ok(OracleStatusResult {
        status: status_str.to_string(),
    })
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OracleResponseResult {
    pub data: serde_json::Value,
    pub status_code: u32,
    pub timestamp: u64,
    pub error: Option<String>,
}

#[op2]
#[serde]
pub fn op_oracle_get_response(
    #[string] request_id: String,
    #[state] oracle_service: &Arc<dyn OracleService>,
) -> Result<OracleResponseResult, AnyError> {
    // Get response
    let rt = tokio::runtime::Runtime::new().unwrap();
    let response = rt.block_on(async {
        oracle_service
            .get_response(&request_id)
            .await
            .map_err(|e| AnyError::msg(format!("Failed to get response: {}", e)))
    })?;

    // Parse response data
    let data = serde_json::from_str(&response.data)
        .unwrap_or_else(|_| serde_json::Value::String(response.data.clone()));

    Ok(OracleResponseResult {
        data,
        status_code: response.status_code,
        timestamp: response.timestamp,
        error: response.error,
    })
}

#[derive(Debug, Serialize, Deserialize)]
pub struct OracleCancelResult {
    pub success: bool,
}

#[op2]
#[serde]
pub fn op_oracle_cancel_request(
    #[string] request_id: String,
    #[state] oracle_service: &Arc<dyn OracleService>,
) -> Result<OracleCancelResult, AnyError> {
    // Cancel request
    let rt = tokio::runtime::Runtime::new().unwrap();
    let success = rt.block_on(async {
        oracle_service
            .cancel_request(&request_id)
            .await
            .map_err(|e| AnyError::msg(format!("Failed to cancel request: {}", e)))
    })?;

    Ok(OracleCancelResult { success })
}

// Convenience operations for specific oracle services

#[derive(Debug, Serialize, Deserialize)]
pub struct PriceRequestConfig {
    pub symbol: String,
    pub currency: Option<String>,
    pub sources: Option<Vec<String>>,
    pub requester_id: String,
}

#[op2]
#[serde]
pub fn op_oracle_get_price(
    #[serde] config: PriceRequestConfig,
    #[state] oracle_service: &Arc<dyn OracleService>,
) -> Result<OracleRequestResult, AnyError> {
    // Create price request
    let price_request = PriceRequest {
        symbol: config.symbol,
        currency: config.currency.unwrap_or_else(|| "USD".to_string()),
        sources: config.sources.unwrap_or_default(),
    };

    // Convert to JSON
    let data = serde_json::to_value(price_request)
        .map_err(|e| AnyError::msg(format!("Failed to serialize price request: {}", e)))?;

    // Create oracle request config
    let oracle_config = OracleRequestConfig {
        request_type: "price".to_string(),
        data,
        callback_url: None,
        requester_id: config.requester_id,
    };

    // Submit request
    op_oracle_submit_request(oracle_config, oracle_service)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RandomRequestConfig {
    pub min: Option<u64>,
    pub max: Option<u64>,
    pub count: Option<u32>,
    pub method: Option<String>,
    pub seed: Option<String>,
    pub requester_id: String,
}

#[op2]
#[serde]
pub fn op_oracle_get_random(
    #[serde] config: RandomRequestConfig,
    #[state] oracle_service: &Arc<dyn OracleService>,
) -> Result<OracleRequestResult, AnyError> {
    // Convert method string to enum
    let method = match config.method.as_deref() {
        Some("secure") => RandomMethod::Secure,
        Some("blockchain") => RandomMethod::Blockchain,
        Some("vrf") => RandomMethod::Vrf,
        None => RandomMethod::Secure,
        _ => {
            return Err(AnyError::msg(format!(
                "Unsupported random method: {}",
                config.method.unwrap_or_default()
            )))
        }
    };

    // Create random request
    let random_request = RandomRequest {
        min: config.min.unwrap_or(0),
        max: config.max.unwrap_or(u64::MAX),
        count: config.count.unwrap_or(1),
        method,
        seed: config.seed,
    };

    // Convert to JSON
    let data = serde_json::to_value(random_request)
        .map_err(|e| AnyError::msg(format!("Failed to serialize random request: {}", e)))?;

    // Create oracle request config
    let oracle_config = OracleRequestConfig {
        request_type: "random".to_string(),
        data,
        callback_url: None,
        requester_id: config.requester_id,
    };

    // Submit request
    op_oracle_submit_request(oracle_config, oracle_service)
}
